### PR TITLE
OpenMPI: New version 5.0.0

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

**Comment by giordano** [comment]
Does this break API/ABI (besides the build)?

**Comment by eschnett** [comment]
It's ABI compatible. API compatible as well since it follows the MPI standard.

However, OpenMPI 5 does not support 32-bit platforms any more...

**Comment by imciner2** [comment]
> It's ABI compatible. API compatible as well since it follows the MPI standard.

Not quite, the release notes do list quite a few removals: https://docs.open-mpi.org/en/v5.0.x/news/news-v5.0.x.html. While the major one is legacy code, it still isn't fully API compatible with the 4.x releases.

**Comment by eschnett** [comment]
I see that the C++ API and the MPIR tools have been removed. Is there anything else?

**Comment by eschnett** [comment]
This PR remove support for 32-bit platforms (explicitly unsupported by OpenMPI 5) and FreeBSD (OpenMPI 5 uses pthread calls that are a GNU extension and are not available there).

File: O/OpenMPI/bundled/patches/recursive_mutex_static_init.patch
Comment (giordano) [inline O/OpenMPI/bundled/patches/recursive_mutex_static_init.patch:1]: Can this patch be upstreamed (or is it already)?

Comment (eschnett) [reply O/OpenMPI/bundled/patches/recursive_mutex_static_init.patch:1]: I can try. This patch makes use of `attribute((constructor))` for the initialization functions which is not part of the C or POSIX standards, although it is widely available in practice. I can certainly create a PR and look for a response.

Comment (eschnett) [reply O/OpenMPI/bundled/patches/recursive_mutex_static_init.patch:1]: See https://github.com/open-mpi/ompi/issues/12029 .

Comment (eschnett) [reply O/OpenMPI/bundled/patches/recursive_mutex_static_init.patch:1]: And now also https://github.com/open-mpi/ompi/pull/12034 .

File: O/OpenMPI/build_tarballs.jl [hunk 1]
```diff
@@ -77,9 +85,20 @@ products = [
 
 dependencies = [
     Dependency("CompilerSupportLibraries_jll"),
+    Dependency("Hwloc_jll"),    # compat="2.0.0"
+    Dependency("PMIx_jll"),     # compat="4.2.0"
+    Dependency("Zlib_jll"),
+    Dependency("libevent_jll"), # compat="2.0.21"
+    Dependency("prrte_jll"),    # compat="3.0.0"
```
Comment (vchuravy) [inline O/OpenMPI/build_tarballs.jl:92]: #7606 I am not sure that using the pprte_jll and PMIx_jll were ready to be fully used as dependencies. I recall having to do weird things to make them work.

https://github.com/JuliaParallel/PMIx.jl/blob/b5c6be4fe342bae7fc6ee770e1e5cc9d388ceab1/src/api.jl#L11

**Pull request merged**